### ISSUE TITLE
Slowly increasing memory consumption while using Appsignal Decorators

### ISSUE THREAD

**Issue by patoms** [appsignal/appsignal-elixir#540]
Hello,

We've been encountering some memory issues with Appsignal Decorators for `@decorate transaction(:my_function)`. When decorating a function that gets called within [`handle_message/3`](https://hexdocs.pm/broadway/Broadway.html#c:handle_message/3) of a Broadway pipeline (long running processes), we see that memory usage of `Appsignal.Transaction.Receiver` is slowly but steadily increasing over time, and it doesn't seem to release any of that memory, unless there is a new deploy. Note that the aforementioned function is called around 1.100.000 times per day.
We're using AppSignal for Elixir package version 1.11.5. Elixir version: :1.9.4/ OTP 22.2.3, dockerized and running on Mesos.

<img width="1394" alt="Screen Shot 2020-02-13 at 4 31 05 PM" src="https://user-images.githubusercontent.com/10545800/74775886-6a707080-524b-11ea-9037-80a46b7f0850.png">

By opening a remote shell/remote observer we saw, that `Appsignal.Transaction.Receiver` consumes most memory, further we tried to gather more information, maybe this helps.

```bash
iex(debug@127.0.0.2)3> Process.whereis(Appsignal.Transaction.Receiver) |> Process.info(:memory)
{:memory, 125575500}
iex(debug@127.0.0.2)4> Process.whereis(Appsignal.Transaction.Receiver) |> Process.info(:garbage_collection)
{:garbage_collection,
 [
   max_heap_size: %{error_logger: true, kill: true, size: 0},
   min_bin_vheap_size: 46422,
   min_heap_size: 233,
   fullsweep_after: 65535,
   minor_gcs: 24953
 ]}
iex(debug@127.0.0.2)5> Process.whereis(Appsignal.Transaction.Receiver) |> Process.info(:garbage_collection_info)
{:garbage_collection_info,
 [
   old_heap_block_size: 987,
   heap_block_size: 610,
   mbuf_size: 6,
   recent_size: 0,
   stack_size: 11,
   old_heap_size: 222,
   heap_size: 249,
   bin_vheap_size: 6,
   bin_vheap_block_size: 46422,
   bin_old_vheap_size: 70,
   bin_old_vheap_block_size: 46422
 ]}
iex(debug@127.0.0.2)6> Process.whereis(Appsignal.Transaction.Receiver) |> Process.info(:binary)
{:binary,
 [
   {140603817535440, 13, 1},
   {140603968540568, 13, 1},
   {140603968540232, 13, 1},
   {140603858590768, 13, 1},
   {140603554977648, 13, 1},
   {140603701354272, 13, 1},
   {140603554160928, 13, 1},
   {140603803601152, 13, 1},
   {140603857087560, 13, 1},
   {140603701407376, 13, 1},
   {140603962403368, 13, 1},
   {140603961288664, 13, 1},
   {140603961616408, 13, 1},
   {140604054195640, 13, 1},
   {140603961092168, 13, 1},
   {140603962374112, 13, 1},
   {140604054189208, 13, 1},
   {140603961627600, 13, 2},
   {140603961286920, 13, 1},
   {140604054202456, 13, 2}
 ]}
iex(debug@127.0.0.2)7> Process.whereis(Appsignal.Transaction.Receiver) |> Process.info()
[
  registered_name: Appsignal.Transaction.Receiver,
  current_function: {Appsignal.Transaction.Receiver, :receiver, 0},
  initial_call: {:proc_lib, :init_p, 5},
  status: :waiting,
  message_queue_len: 0,
  links: [#PID<0.2179.0>],
  dictionary: [
    "$initial_call": {Appsignal.Transaction.Receiver, :"-start_link/0-fun-0-",
     0},
    "$ancestors": [Appsignal.Supervisor, #PID<0.2177.0>],
    "$callers": [#PID<0.2179.0>]
  ],
  trap_exit: false,
  error_handler: :error_handler,
  priority: :normal,
  group_leader: #PID<0.2176.0>,
  total_heap_size: 2591,
  heap_size: 1598,
  stack_size: 11,
  reductions: 152277192,
  garbage_collection: [
    max_heap_size: %{error_logger: true, kill: true, size: 0},
    min_bin_vheap_size: 46422,
    min_heap_size: 233,
    fullsweep_after: 65535,
    minor_gcs: 24955
  ],
  suspending: []
]
iex(debug@127.0.0.2)8> Process.whereis(Appsignal.Transaction.Receiver) |> :erlang.garbage_collect()
true
iex(debug@127.0.0.2)9> Process.whereis(Appsignal.Transaction.Receiver) |> Process.info(:memory)
{:memory, 125559624}
```

### Usage
```elixir
  @decorate transaction(:my_function)	
  def my_function(param1, param2) do
     # some implementation
  end
```

**Comment by whatyouhide** [comment]
FWIW, we have observed this across several projects that use the latest Appsignal versions. On other projects, this issue doesn't appear to be happening. One of those in particular is not suffering from this and running appsignal-elixir 1.10.12. We would love to help debug this but the problem is that the memory leak shows over several hours so it's not straightforward to bisect where the thing that introduced it appeared. Thanks in advance folks!

**Comment by jeffkreeftmeijer** [comment]
Hey @patoms and @whatyouhide,

Thanks for your detailed description of this problem. We’re working to find where this issue was introduced, and we’ll keep this issue updated with our findings.

In the meantime, our testing seems to show that 1.10.12 doesn’t produce the same results. @patoms (and others possibly running into this issue), could you downgrade to that version and let us know if that resolves the issue for you?

Thanks!

**Comment by jeffkreeftmeijer** [comment]
Hey both,

We found an issue that matches your descriptions and could cause memory bloat caused by lingering monitors in high-throughput apps that run multiple transactions from a single process. I've pushed a patch to the [receiver-references](https://github.com/appsignal/appsignal-elixir/tree/receiver-references) branch.

We're doing thorough testing before releasing this patch as 1.11.6, but it would be great if you could run this branch in your apps, and report your findings here. The branch contains 1.11.15, plus [the patch](https://github.com/appsignal/appsignal-elixir/commit/6384f42cd2bb21d9dc7ef4956cc18b3ae61eef5e).

      {:appsignal, github: "appsignal/appsignal-elixir", branch: "receiver-references"}

Thanks again!

**Comment by patoms** [comment]
> Hey both,
> 
> We found an issue that matches your descriptions and could cause memory bloat caused by lingering monitors in high-throughput apps that run multiple transactions from a single process. I've pushed a patch to the [receiver-references](https://github.com/appsignal/appsignal-elixir/tree/receiver-references) branch.
> 
> We're doing thorough testing before releasing this patch as 1.11.6, but it would be great if you could run this branch in your apps, and report your findings here. The branch contains 1.11.15, plus [the patch](https://github.com/appsignal/appsignal-elixir/commit/6384f42cd2bb21d9dc7ef4956cc18b3ae61eef5e).
> 
> ```
>   {:appsignal, github: "appsignal/appsignal-elixir", branch: "receiver-references"}
> ```
> 
> Thanks again!

Awesome! Thank you, @jeffkreeftmeijer. We'll deploy this version today and get back to you with results!

cc @whatyouhide

**Comment by jeffkreeftmeijer** [comment]
We've released #541 as 1.12.6, which should fix this issue. Could you upgrade to that version and reopen this issue if the problem persists?

Thanks!

**Comment by patoms** [comment]
> We've released #541 as 1.11.6, which should fix this issue. Could you upgrade to that version and reopen this issue if the problem persists?
> 
> Thanks!

Awesome! Thank you @jeffkreeftmeijer. We've tested with the branch you mentioned above and memory seems stable/we didn't see the increasing memory consumption anymore! Happy to upgrade now :)

<img width="1411" alt="Screen Shot 2020-02-21 at 9 53 00 AM" src="https://user-images.githubusercontent.com/10545800/75058505-f83c9d80-548f-11ea-848b-ae26804a96e8.png">
